### PR TITLE
Insert filecache data and conflict fallback read in the same transaction

### LINKED ISSUE #21811: RuntimeException: File entry could not be inserted but could also not be selected with getId() in order to perform an update
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps to reproduce
1. Open folder on Web or try to sync it
2. On the top right corner an error appears and on Sync Client it stops syncing
3.

### Expected behaviour
The folder should open and the content should be accessible

### Actual behaviour
- An error occurs and the users gets thrown back to the main page
- On the sync client the user gets an error message, that the file is not accessible
- This doesn't happen on every file or folder, only on a few (I couldn't make out a pattern)

### Server configuration

**Operating system:**
Ubuntu 20.04
**Web server:**
Apache 2.4.29
**Database:**
PostgreSQL 12.2
**PHP version:**
7.4.3
**Nextcloud version:** (see Nextcloud admin page)
19.0.0
**Updated from an older Nextcloud/ownCloud or fresh install:**
No
**Where did you install Nextcloud from:**
Nextcloud Hansson VM
**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.5.0
  - activity: 2.12.0
  - admin_audit: 1.9.0
  - cloud_federation_api: 1.2.0
  - comments: 1.9.0
  - contactsinteraction: 1.0.0
  - dav: 1.15.0
  - extract: 1.2.4
  - federatedfilesharing: 1.9.0
  - federation: 1.9.0
  - files: 1.14.0
  - files_external: 1.10.0
  - files_inotify: 0.1.8
  - files_pdfviewer: 1.8.0
  - files_rightclick: 0.16.0
  - files_sharing: 1.11.0
  - files_trashbin: 1.9.0
  - files_versions: 1.12.0
  - files_videoplayer: 1.8.0
  - firstrunwizard: 2.8.0
  - groupfolders: 6.0.6
  - impersonate: 1.6.1
  - logreader: 2.4.0
  - lookup_server_connector: 1.7.0
  - nextcloud_announcements: 1.8.0
  - notifications: 2.7.0
  - oauth2: 1.7.0
  - password_policy: 1.9.1
  - photos: 1.1.0
  - privacy: 1.3.0
  - provisioning_api: 1.9.0
  - recommendations: 0.7.0
  - serverinfo: 1.9.0
  - settings: 1.1.0
  - sharebymail: 1.9.0
  - support: 1.2.1
  - survey_client: 1.7.0
  - systemtags: 1.9.0
  - text: 3.0.1
  - theming: 1.10.0
  - twofactor_backupcodes: 1.8.0
  - updatenotification: 1.9.0
  - user_ldap: 1.9.0
  - viewer: 1.3.0
  - workflowengine: 2.1.0
Disabled:
  - calendar
  - contacts
  - deck
  - encryption
  - mail
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "xxx.xxx.xxx.xxx",
            "nc.example.com",
            "nc.example.com"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "pgsql",
        "version": "19.0.0.12",
        "overwrite.cli.url": "https:\/\/nc.example.com\/",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "upgrade.disable-web": "true",
        "log_type": "file",
        "logfile": "\/var\/log\/nextcloud\/nextcloud.log",
        "loglevel": "2",
        "log.condition": {
            "apps": [
                "admin_audit"
            ]
        },
        "mail_smtpmode": "smtp",
        "remember_login_cookie_lifetime": "1800",
        "log_rotate_size": "10485760",
        "trashbin_retention_obligation": "auto, 180",
        "versions_retention_obligation": "auto, 365",
        "simpleSignUpLink.shown": "false",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "filelocking.enabled": true,
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "timeout": 0.5,
            "dbindex": 0,
            "password": "***REMOVED SENSITIVE VALUE***"
        },
        "logtimezone": "Europe\/Zurich",
        "htaccess.RewriteBase": "\/",
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "OCA\\User_LDAP\\LDAPProviderFactory",
        "maintenance": false,
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_sendmailmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "skeletondirectory": "\/var\/www\/nextcloud\/core\/skeleton"
    }
}
```
</details>

**Are you using external storage, if yes which one:** SMB external storage

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** ActiveDirectory

#### LDAP configuration (delete this part if not used)
<details>
<summary>LDAP config</summary>

```
{
    "apps": {
        "user_ldap": {
            "background_sync_prefix": "s01",
            "installed_version": "1.9.0",
            "s01ldap_group_filter_mode": "0",
            "s01ldap_gid_number": "gidNumber",
            "s01ldap_user_display_name_2": "",
            "s01ldap_group_display_name": "cn",
            "s01ldap_tls": "0",
            "s01ldap_quota_def": "",
            "s01ldap_quota_attr": "",
            "s01ldap_email_attr": "",
            "s01ldap_cache_ttl": "600",
            "s01home_folder_naming_rule": "",
            "s01ldap_turn_off_cert_check": "0",
            "s01ldap_attributes_for_user_search": "",
            "s01ldap_attributes_for_group_search": "",
            "s01ldap_expert_username_attr": "",
            "s01ldap_expert_uuid_group_attr": "",
            "s01use_memberof_to_detect_membership": "1",
            "s01last_jpegPhoto_lookup": "0",
            "s01ldap_nested_groups": "0",
            "s01ldap_paging_size": "500",
            "s01ldap_turn_on_pwd_change": "0",
            "s01ldap_experienced_admin": "0",
            "s01ldap_dynamic_group_member_url": "",
            "s01ldap_default_ppolicy_dn": "",
            "s01ldap_user_avatar_rule": "default",
            "s01ldap_ext_storage_home_attribute": "",
            "s01ldap_userlist_filter": "(&(|(objectclass=user))(|(|(memberof=CN=example.sg_nextcloud_test,OU=example,OU=example,OU=example,OU=example,DC=example,DC=local)(primaryGroupID=3867))))",
            "s01ldap_base_users": "DC=example,DC=local",
            "s01ldap_dn": "CN=Nextcloud LDAP,OU=example,DC=example,DC=local",
            "s01ldap_groupfilter_objectclass": "group",
            "s01ldap_agent_password": "***REMOVED SENSITIVE VALUE***",
            "s01ldap_port": "389",
            "s01ldap_loginfilter_attributes": "sAMAccountName",
            "s01ldap_host": "ldap:\/\/example.example.local",
            "s01ldap_base_groups": "DC=example,DC=local",
            "s01ldap_groupfilter_groups": "example.sg_nextcloud_test",
            "s01has_memberof_filter_support": "1",
            "s01ldap_loginfilter_email": "1",
            "s01ldap_login_filter": "(&(&(|(objectclass=user))(|(|(memberof=CN=example.sg_nextcloud_test,OU=example,OU=example,OU=example,OU=example,DC=example,DC=local)(primaryGroupID=3867))))(|(samaccountname=%uid)(|(mailPrimaryAddress=%uid)(mail=%uid))(|(sAMAccountName=%uid))))",
            "s01ldap_configuration_active": "1",
            "s01ldap_userfilter_groups": "example.sg_nextcloud_test",
            "s01ldap_group_filter": "(&(|(objectclass=group))(|(cn=example.sg_nextcloud_test)))",
            "s01ldap_group_member_assoc_attribute": "member",
            "s01ldap_display_name": "displayname",
            "types": "authentication",
            "enabled": "yes",
            "s01ldap_userfilter_objectclass": "user",
            "background_sync_offset": "0",
            "background_sync_interval": "43200",
            "s01ldap_user_filter_mode": "1",
            "s01ldap_base": "DC=example,DC=local",
            "s01ldap_backup_host": "",
            "s01ldap_backup_port": "",
            "s01ldap_override_main_server": "",
            "s01ldap_login_filter_mode": "0",
            "s01ldap_loginfilter_username": "1",
            "s01ldap_expert_uuid_user_attr": "samAccountName",
            "cleanUpJobOffset": "0",
            "s01_lastChange": "1594639533"
        }
    }
}

```
</details>

### Client configuration
**Browser:**
Firefox 78.0.2 / Sync Client Windows Version 2.6.2 stable
**Operating system:**
Windows 10 1909
#### Web server error log
<details>
<summary>Web server error log</summary>

```
No directly related Apache errors, only this:
[Mon Jul 13 13:24:01.932280 2020] [access_compat:error] [pid 1144859:tid 139691719460608] [client xxx.xxx.xxx.xxx:52032] AH01797: client denied by server configuration: /var/www/nextcloud/config
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
{
   "reqId":"HYuCZrMDXF1h6dlMxIou",
   "level":4,
   "time":"2020-07-13T13:35:36+02:00",
   "remoteAddr":"195.65.36.7",
   "user":"cba.testhesi",
   "app":"webdav",
   "method":"PROPFIND",
   "url":"/remote.php/dav/files/USER/FOLDER/PATH",
   "message":{
      "Exception":"RuntimeException",
      "Message":"File entry could not be inserted but could also not be selected with getId() in order to perform an update. Please try again.",
      "Code":0,
      "Trace":[
         {
            "file":"/var/www/nextcloud/lib/private/Files/Cache/Wrapper/CacheWrapper.php",
            "line":131,
            "function":"insert",
            "class":"OC\\Files\\Cache\\Cache",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/lib/private/Files/Cache/Scanner.php",
            "line":294,
            "function":"insert",
            "class":"OC\\Files\\Cache\\Wrapper\\CacheWrapper",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/lib/private/Files/Cache/Scanner.php",
            "line":224,
            "function":"addToCache",
            "class":"OC\\Files\\Cache\\Scanner",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/lib/private/Files/Cache/Scanner.php",
            "line":423,
            "function":"scanFile",
            "class":"OC\\Files\\Cache\\Scanner",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/lib/private/Files/Cache/Scanner.php",
            "line":388,
            "function":"handleChildren",
            "class":"OC\\Files\\Cache\\Scanner",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/lib/private/Files/Cache/Scanner.php",
            "line":340,
            "function":"scanChildren",
            "class":"OC\\Files\\Cache\\Scanner",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/lib/private/Files/View.php",
            "line":1339,
            "function":"scan",
            "class":"OC\\Files\\Cache\\Scanner",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/lib/private/Files/View.php",
            "line":1383,
            "function":"getCacheEntry",
            "class":"OC\\Files\\View",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/apps/dav/lib/Connector/Sabre/Directory.php",
            "line":221,
            "function":"getFileInfo",
            "class":"OC\\Files\\View",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Tree.php",
            "line":80,
            "function":"getChild",
            "class":"OCA\\DAV\\Connector\\Sabre\\Directory",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php",
            "line":976,
            "function":"getNodeForPath",
            "class":"Sabre\\DAV\\Tree",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php",
            "line":1678,
            "function":"getPropertiesIteratorForPath",
            "class":"Sabre\\DAV\\Server",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php",
            "line":1661,
            "function":"writeMultiStatus",
            "class":"Sabre\\DAV\\Server",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php",
            "line":363,
            "function":"generateMultiStatus",
            "class":"Sabre\\DAV\\Server",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php",
            "line":89,
            "function":"httpPropFind",
            "class":"Sabre\\DAV\\CorePlugin",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php",
            "line":474,
            "function":"emit",
            "class":"Sabre\\DAV\\Server",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php",
            "line":251,
            "function":"invokeMethod",
            "class":"Sabre\\DAV\\Server",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php",
            "line":319,
            "function":"start",
            "class":"Sabre\\DAV\\Server",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/apps/dav/lib/Server.php",
            "line":320,
            "function":"exec",
            "class":"Sabre\\DAV\\Server",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/apps/dav/appinfo/v2/remote.php",
            "line":35,
            "function":"exec",
            "class":"OCA\\DAV\\Server",
            "type":"->"
         },
         {
            "file":"/var/www/nextcloud/remote.php",
            "line":167,
            "args":[
               "/var/www/nextcloud/apps/dav/appinfo/v2/remote.php"
            ],
            "function":"require_once"
         }
      ],
      "File":"/var/www/nextcloud/lib/private/Files/Cache/Cache.php",
      "Line":324,
      "CustomMessage":"--"
   },
   "userAgent":"Mozilla/5.0 (Windows) mirall/2.6.2stable-Win64 (build 20191224) (Nextcloud)",
   "version":"19.0.0.12",
   "id":"5f0c47095f5de"
}
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
Insert your browser log here, this could for example include:


scheme | https
-- | --
host | nc.example.com
filename | /ocs/v2.php/apps/text/workspace
 
path | FOLDER/PATH
 
Adresse | xxx.xxx.xxx.xxx:443


```
</details>

### PR DESCRIPTION
**Pull request by tcitworld**
* Resolves: #21811

## Summary

In the case of an asynchronous replication database system, the inserted row might not yet be available in the replicated R/O database.

This removes d9184584e08def794a298966afde24f232b7abf5 from @icewind1991 which shouldn't be needed anymore.

Best reviewed as https://github.com/nextcloud/server/pull/36313/files?diff=unified&w=1

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/private/Files/Cache/Cache.php [hunk 1]
```diff
@@ -271,7 +276,8 @@ public function put($file, array $data) {
 	 * @param array $data
 	 *
 	 * @return int file id
-	 * @throws \RuntimeException
+	 * @throws Exception
```
Comment (come-nc) [inline lib/private/Files/Cache/Cache.php:279]:
```suggestion
	 * @throws \Exception
```

**Comment by come-nc** [comment]
I’m not comfortable enough with the DB code to review that, I never used transactions

**Comment by tcitworld** [comment]
Needs investigation for the PostgreSQL failing tests.

**Comment by flohoff** [comment]
We are running with a rebased version of this in 27.1.9 now and the getId backtraces disappeared completely. We saw them for years now. Backend is a Galera Cluster with a Maxscale.

**Pull request closed without merging**